ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,777 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * GetBlobOperation class is responsible for maintaining the state associated with a GetBlob operation, and completing
+ * it. It would help to look at this class in conjunction with {@link PutOperation}.
+ *
+ * Blobs that need to be fetched are either simple (those that consist of only a single chunk) or composite (those
+ * that consists of one metadata chunk and two or more data chunks). Whether a blob is simple or composite can only be
+ * determined after the first chunk is fetched, based on whether the first chunk turns out to be a metadata chunk or
+ * a data chunk.
+ *
+ * As soon as the first data chunk is fetched (which could be the first chunk in the case of simple blobs and the
+ * second chunk fetched in the case of composite blobs), the operation callback is invoked (and the future is marked
+ * as done) so that the caller can start reading in data. The rest of the chunks are asynchronously fetched and
+ * buffered up to the maximum that can be buffered. When fetched chunks are consumed by the caller, subsequent chunks
+ * become eligible to be fetched.
+ *
+ * This class implements {@link ReadableStreamChannel} and passes itself as the result of this operation.
+ */
+class GetBlobOperation extends GetOperation<ReadableStreamChannel> implements ReadableStreamChannel {
+  // the callback to use to complete the operation.
+  private final OperationCompleteCallback operationCompleteCallback;
+  // whether the operationCompleteCallback has been called already.
+  private final AtomicBoolean operationCallbackInvoked = new AtomicBoolean(false);
+  // The first chunk may be a metadata chunk if the blob is composite, or the only data chunk if the blob is simple.
+  private final FirstGetChunk firstChunk;
+  // Associated with all data chunks in the case of composite blobs. Only a fixed number of these are initialized.
+  // Each of these is initialized with the information required to fetch a data chunk and is responsible for
+  // retrieving and adding it to the list of chunk buffers. Once complete, they are reused to fetch subsequent data
+  // chunks.
+  private GetChunk[] dataChunks;
+  // the factory to use to deserialize keys in a metadata chunk.
+  private final BlobIdFactory blobIdFactory;
+  // the total number of data chunks associated with this blob.
+  private int numChunksTotal;
+  // the total number of data chunks retrieved so far (and may or may not have been written out yet).
+  private int numChunksRetrieved;
+  // a list iterator to the chunk ids that need to be fetched for this operation, if this is a composite blob.
+  private ListIterator<StoreKey> chunkIdIterator;
```

Comment:
in case the above comment wasn't clear - a call has to be made to get both the next chunk and that chunk's index. We eventually do that by making calls to `listIterator.nextIndex()` and `listIterator.next()` and they have to be done in that order as calling `listIterator.next()` advances the cursor.
This is okay if the call is made using the `listIterator`, as this is its behavior. For `firstChunk`, this may not be immediately clear.